e2e-test-server/src/main/java/com/google/cloud/opentelemetry/endtoend/PubSubPushServer.java

```diff
@@ -0,0 +1,221 @@
+ * like a daemon when no request is processing). This means we need a dedicated server actively
+ * listening on a port instead of waiting on messages being pushed to a topic which can then be
+ * pulled (this is what happens in 'pull' subscription mode). This server sets up an {@link
+ * HttpServer} on a specified port which actively listens for incoming requests.
+ *
+ * <p>More information on why 'push' mode is preferred can be seen <a
+ * href="https://cloud.google.com/run/docs/triggering/pubsub-push#:~:text=Note%3A%20Google,minimum%20instances">here</a>.
+ *
+ * <p>This class is responsible for the following:
+ *
+ * <ul>
+ *   <li>Setting up a {@link HttpServer} bound to a specified port, listening for incoming requests.
+ *   <li>Validate incoming HTTP requests and parse them into {@link PubsubMessage}s.
+ *   <li>Hand off the parsed {@link PubsubMessage}s to a specified {@link PubSubMessageHandler}.
+ * </ul>
+ */
+public class PubSubPushServer implements PubSubServer {
+
+  private static final String POST_REQUEST = "POST";
+
+  private final int port;
+  private final HttpServer httpServer;
+  private final HttpHandler rootRequestHandler;
+  private final PubSubMessageHandler pubsubMessageHandler;
+
+  /**
+   * Public constructor for the {@link PubSubPushServer}.
+   *
+   * @param port The port on which the HTTP server should listen for incoming requests.
+   * @param pubSubMessageHandler The {@link PubSubMessageHandler} responsible for handling the
+   *     incoming HTTP requests.
+   */
+  public PubSubPushServer(int port, PubSubMessageHandler pubSubMessageHandler) {
+    this.port = port;
+    this.pubsubMessageHandler = pubSubMessageHandler;
+    this.rootRequestHandler = createRootRequestHandler();
+    this.httpServer = createHttpServer();
+  }
+
+  @Override
+  public void start() {
+    this.httpServer.start();
+  }
+
+  @Override
+  public void close() {
+    httpServer.stop(60);
+    pubsubMessageHandler.cleanupMessageHandler();
+  }
+
+  private PubsubMessage parseIncomingMessage(HttpExchange httpExchange) {
+    InputStreamReader inputStreamReader =
+        new InputStreamReader(httpExchange.getRequestBody(), StandardCharsets.UTF_8);
+    JsonElement jsonRoot = JsonParser.parseReader(inputStreamReader);
+    String msgStr = jsonRoot.getAsJsonObject().get("message").toString();
+    Gson gson = new Gson();
+    Message message = gson.fromJson(msgStr, Message.class);
+    return PubsubMessage.newBuilder().putAllAttributes(message.getAttributes()).build();
+  }
+
+  private HttpServer createHttpServer() {
+    HttpServer httpServer = null;
+    try {
+      httpServer = HttpServer.create(new InetSocketAddress(this.port), 0);
+      httpServer.createContext("/", this.rootRequestHandler);
+      httpServer.setExecutor(MoreExecutors.directExecutor());
+    } catch (IOException e) {
+      e.printStackTrace();
+    }
+    return httpServer;
+  }
+
+  private HttpHandler createRootRequestHandler() {
```

Comment:
I would not be opposed to using some minimal framework to avoid a lot of this boilerplate. E.g. https://sparkjava.com/ but not sure if there's a better option.

Reply:
I feel its actually better to use some framework to excess code. Did not use it right now since I was not sure how we feel about introducing additional dependencies to 3rd party libraries. I can address this in a follow-up PR.